File: [89] Gray Code/89.gray-code.cpp
/*
 * @lc app=leetcode id=89 lang=cpp
 *
 * [89] Gray Code
 */

// @lc code=start
#include <vector>
#include <unordered_set>
using namespace std;
class Solution {
public:
    vector<int> grayCode(int n) {
        // backtracking DFS
        vector<int> res;
        res.push_back(0);
        unordered_set<int> isPresent;
        isPresent.insert(0);
        DFS(res, n, isPresent);
        return res;
    }

private:
    bool DFS(vector<int>& res, int n, unordered_set<int>& isPresent) {
        if (res.size() == (1 << n)) return true;
        int current = res.back();
        for (int i = 0; i < n; i++) {
            int next = current ^ (1 << i);
            if (isPresent.find(next) == isPresent.end()) {
                res.push_back(next);
                isPresent.insert(next);
                if(DFS(res, n, isPresent)) return true;
                res.pop_back();
                isPresent.erase(next);
            }
        }
        return false;
    }
};
// @lc code=end

File: [89] Gray Code/89.gray-code.java
import java.util.ArrayList;

/*
 * @lc app=leetcode id=89 lang=java
 *
 * [89] Gray Code
 */

// @lc code=start
class Solution {
    public List<Integer> grayCode(int n) {
        // Recursion
        // List<Integer> res = new ArrayList<>();
        // recurse(res, n);
        // return res;
        // In fact it can be rewrote as iteration method
        List<Integer> res = new ArrayList<>();
        res.add(0);
        for (int i = 1; i <= n; i++) {
            int res_len = res.size();
            int mask = 1 << (i-1);
            for (int j = res_len-1; j >= 0; j--) {
                res.add(res.get(j) | mask);
            }
        }
        return res;
        // The most elegant solution is to find the relationship between gray(i) and i
        // gray(i) = i XOR (i>>1)
        // int seq_len = 1 << n;
        // for (int i = 0; i < seq_len; i++) {
        //     res.add(i^(i>>1));
        // }
        // return res;
    }

    // private void recurse(List<Integer> res, int n) {
    //     if (n == 0) {
    //         res.add(0);
    //         return;
    //     } else {
    //         recurse(res, n-1);
    //         int res_len = res.size();
    //         int mask = 1 << (n-1);
    //         for (int i = res_len-1; i >= 0; i--) {
    //             res.add(res.get(i) | mask);
    //         }
    //     }
    // }
}
// @lc code=end

